docs/sdk/src/reference_docs/defensive_programming.rs

```diff
@@ -0,0 +1,618 @@
+//!
+//! ## Defensive Programming
+//!
+//! [Defensive programming](https://en.wikipedia.org/wiki/Defensive_programming) is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior, input or events which may arise in runtime. Normally,
+//! unforeseen circumstances may cause the program to stop or, in the Rust context, `panic!`.
+//! Defensive practices allow for these circumstances to be accounted for ahead of time and for them
+//! to be handled in a graceful manner, which is in the line of the intended, fault-tolerant
+//! behavior of blockchains.
+//!
+//! The Polkadot SDK is both built to reflect these principles and to facilitate their usage
+//! accordingly.
+//!
+//! ## General Practices
+//!
+//! When developing within the context of the a runtime, there is *one* golden rule:
+//!
+//! ***DO NOT PANIC***. There are some exceptions, such as critical operations being actually more
+//! dangerous than allowing the runtime to continue functioning (block authoring, consensus, etc).
+//!
+//! > It's important to make the differentiation between the **runtime** and **node**.  The runtime
+//! > refers to the core business logic of a Substrate-based chain, whereas the node refers to the
+//! > outer client which deals with telemetry and gossip from other nodes. For more information,
+//! > read about Substrate's architecture.
+//! > It's also important to note that the behavior of the **node** may differ from that of the
+//! > **runtime**, which is also why at times, you may see `unwrap()` or other "non-defensive"
+//! > behavior taking place.
+//!
+//!  General guidelines:
+//!
+//! - Avoid writing functions that could explicitly panic. Directly using `unwrap()` for a
+//!   [`Result`], or common errors such as accessing an out of bounds index on a collection should
+//!   not be used. Safer methods to access collection types, i.e., `get()` are available, upon which
+//!   defensive handling of the resulting [`Option`] can occur.
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.  Even this is
+//!   discouraged, however, as future changes to that function could then cause that statement to
+//!   panic.  It is better to ensure all errors are propagated and handled accordingly in some way.
+//! - If a function *can* panic, it usually is prefaced with `unchecked_` to indicate its unsafety.
+//! - If you are writing a function that could panic, [be sure to document it!](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components)
+//! - Carefully handle mathematical operations.  Many seemingly, simplistic operations, such as
+//!   **arithmetic** in the runtime, could present a number of issues [(see more later in this
+//!   document)](#integer-overflow).
+//!
+//! ### Examples of when to `panic!`
```

Comment:
I would hesitate mentioning this toward the beginning of this article, as it could be misleading. 

95% of pallet developers are not going to write pallets that are consensus level, and are therefore never in need of having an explicit panic. So I would first address the majority, and leave this as some kind of appendix.

Reply:
I placed it at the end of the doc and specified this point.